Update data-pipelines.md (#1645)
iterative/dvc.org: content/docs/start/data-pipelines.md

DVC pipelines (`dvc.yaml` file, `dvc run`, and `dvc repro` commands) solve a few
important problems:

- _Automation_ - run a sequence of steps in a "smart" way that makes iterating on
  your project faster. DVC automatically determines which parts of a project
  need to be run, and it caches "runs" and their results, to avoid unnecessary
  re-runs.
- _Reproducibility_ - `dvc.yaml` and `dvc.lock` files describe what data to use
  and which commands will generate the pipeline results (such as an ML
  model). Storing these
  files in Git makes it easy to version and share.
- _Continuous Delivery and Continuous Integration (CI/CD) for ML_ - describing

>   projects in way that it can be reproduced (built) is the fist necessary step

Reviewer: it -> they?

Reviewer 2: Oops I merged the REstyled PR #1647. I'll continue this elsewhere... ⏳

DVC pipelines (`dvc.yaml` file, `dvc run`, and `dvc repro` commands) solve a few
important problems:

- _Automation_ - run a sequence of steps in a "smart" way to iterate on your 
  project faster. DVC caches "runs" and results in stages to avoid unnecessary

>   re-runs.

Reviewer 2: - "makes iterating on the project faster" was more accurate I think.
- "It automatically determines which parts of a project need to be run" was very important"

- _Reproducibility_ - YAML files describe and capture what data to use and
  what commands to run to produce an ML model. Storing these files in Git

>   makes it easy to version and share.

Reviewer 2: Let's use `` `dvc.yaml` and `dvc.lock` files describe and capture what data...``

I agree with the other changes to this bullet 👍

- _Continuous Delivery and Continuous Integration (CI/CD) for ML_ - describing

>   reproducible ML pipelines (builds) facilitates CI/CD systems.

Reviewer 2: Definitely better. But not sure about "(builds)" What builds? Builds what? "builds" the plural noun? No need to summarize so much that it's ambiguous, let's try to be as explicit as necessary. Same with "facilitates": how does it facilitate them?

Author: Thank you for the feedback! I agree. Hopefully my additions are more explicit.

DVC pipelines (`dvc.yaml` file, `dvc run`, and `dvc repro` commands) solve a few
important problems:

> - _Automation_ - run a sequence of steps in a "smart" way to make iterating on

Reviewer 2: ```suggestion
- _Automation_ - run a sequence of steps in a "smart" way that makes iterating on
```
Per https://github.com/iterative/dvc.org/pull/1645#discussion_r463785986

>   your project faster. DVC caches "runs" and results in stages and automatically
>   determines which parts of a project need to be run to avoid unnecessary
>   re-runs.

Reviewer 2: ```suggestion
  your project faster. DVC automatically determines which parts of a project
  need to be run, and caches "runs" and their results, to avoid unnecessary
  re-runs.
```
Also per https://github.com/iterative/dvc.org/pull/1645#discussion_r463785986

Reviewer: are we merging this change? ^^

Reviewer 2: Oh yeah forgot to commit this... Thanks

- _Reproducibility_ - `dvc.yaml` and `dvc.lock` files describe and capture what

>   data to use and what commands to run to produce an ML model. Storing these

Reviewer 2: > to produce an ML model

It used to be "model, for example". That's because data pipelines can have any results, it doesn't even have to be an ML application. Would be ideal to have it more general (but still mention the "ML model" key word).

Reviewer 2: Maybe something like

...files describe what data to use and what commands to execute to generate results (such as ML models).

Author: ah yes I am definitely tunnel visioned toward my use case. Pipeline results is nice and general 👍

Reviewer 2: OK, I'll update it.

> - _Reproducibility_ - `dvc.yaml` and `dvc.lock` files describe and capture what
>   data to use and what commands to run to produce an ML model. Storing these

Reviewer 2: ```suggestion
- _Reproducibility_ - `dvc.yaml` and `dvc.lock` files describe what data to use
  and which commands will generate the pipeline results (such as an ML
  model). Storing these
```

  files in Git makes it easy to version and share.
- _Continuous Delivery and Continuous Integration (CI/CD) for ML_ - reproducible
  ML pipelines allow CI/CD systems to retrain models on fresh

>   datasets with identical training, and save the results.

Reviewer 2: ```suggestion
  datasets with identical training, save the results, and even produce reports
  about the whole process. See [CML.dev](https://cml.dev/) for some examples.
```
Per https://github.com/iterative/dvc.org/pull/1645#pullrequestreview-459404692

DVC pipelines (`dvc.yaml` file, `dvc run`, and `dvc repro` commands) solve a few
important problems:

- _Automation_ - run a sequence of steps in a "smart" way that makes iterating on
  your project faster. DVC caches "runs" and results in stages and automatically
  determines which parts of a project need to be run to avoid unnecessary
  re-runs.
- _Reproducibility_ - `dvc.yaml` and `dvc.lock` files describe what data to use
  and which commands will generate the pipeline results (such as an ML
  model). Storing these
  files in Git makes it easy to version and share.
- _Continuous Delivery and Continuous Integration (CI/CD) for ML_ - reproducible

>   ML pipelines allow CI/CD systems to retrain models on fresh

Reviewer: this describes a very specific use case. CI/CD can include running the last stage for example, full retraining, just training remotely instead of local ... etc

> - _Continuous Delivery and Continuous Integration (CI/CD) for ML_ - reproducible
>   ML pipelines allow CI/CD systems to retrain models on fresh
>   datasets with identical training, save the results, and even produce reports
>   about the whole process. See [CML.dev](https://cml.dev/) for some examples.

Reviewer 2: How about this @shcheklein ?
```suggestion
- _Continuous Delivery and Continuous Integration (CI/CD) for ML_ - reproducible
  ML pipelines allow CI/CD systems to reproduce all or part of the process (e.g.
  model training) on production datasets. Using DVC, they can then save the
  results, and even produce reports about the whole process. See
  [CML.dev](https://cml.dev/) for some examples.
```

Reviewer 2: Feel free to edit/commit it and merge the Restyled PR if you agree 🙂 or lmk and I'll do so.

Reviewer: it still describes one very specific case "reproduce all or part of the process ... on production datasets"

let's go back a bit, gusy. What was the problem with initial version? So, that I understand how to help you solve it.

Reviewer 2: It just had a small typo, Noah decided to rewrite it all, for readability I think. Here's the original version (typo fixed) in case we want to just go back to that:
```suggestion
- _Continuous Delivery and Continuous Integration (CI/CD) for ML_ - describing
  projects in way that it can be reproduced (built) is the fist necessary step
  before introducing CI/CD systems.
```

Reviewer: So, let's just fix the typo then? if the problem is not clear. Changing language a bit is fine either if it sounds/reads better, but let's not change the essence if there is no clear understanding what exactly are we fixing.

Reviewer 2: Yes, that's what I meant. The typo is fixed in the suggestion I left. We just need to commit it. Allow me...

Author: To me the original version echos reproducibility rather than pipeline flexibility. The piece that matters for CI/CD is DVC can figure out when something important has changed (data, code, etc) and re-run a pipeline correctly. Like what @jorgeorpinel said "pipelines allow CI/CD systems reproduce all or part of the process."

Reviewer 2: Thanks Noah, I'll chat with Ivan and continue this elsewhere if needed, since this has been merged. But feel free to contribute another PR if you prefer 🙂

Author: No problem! Again, thanks for all of the feedback

Reviewer: yep. I agree with the `pipelines allow CI/CD systems reproduce all or part of the process` part. `on production datasets` and mentioning `train` make it specific.

> The piece that matters for CI/CD is DVC can figure out when something important has changed (data, code, etc) and re-run a pipeline correctly

not sure this it true. Since it's already committed it can force `dvc repro` for example to guarantee that external system gets the same result. There are other examples when we can "abuse" CI/CD to do the training first time- we don't change anything per-se - everything comes with a commit, system runs it first time.

Reviewer: Also, I think CML.dev link still make a lot of sense.

DVC pipelines (`dvc.yaml` file, `dvc run`, and `dvc repro` commands) solve a few
important problems:

- _Automation_ - run a sequence of steps in a "smart" way to make iterating on
  your project faster. DVC caches "runs" and results in stages and automatically
  determines which parts of a project need to be run to avoid unnecessary

>   re-runs.

Reviewer 2: - Again, "way that makes iterating" is more accurate.
- `DVC caches "runs" and results in stages` - this isn't accurate. The text here was referring to a feature called run-cache which we haven't documented yet. Stages are not where results are cached.

Reviewer 2: How about something like

DVC automatically determines which parts of a project need to be run, and caches "runs" and their results, to avoid unnecessary re-runs.

Author: That sounds great to me! Thank you for the assist here

Reviewer 2: OK, will apply 🙂

- _Reproducibility_ - `dvc.yaml` and `dvc.lock` files describe and capture what
  data to use and what commands to run to produce an ML model. Storing these
  files in Git makes it easy to version and share.

> - _Continuous Delivery and Continuous Integration (CI/CD) for ML_ - reproducible

Reviewer: Noah, first of all thanks for the contribution! 🙏 

Can we though keep this CI/CD piece a bit simpler? Potentially, replace "forward" links like repro and metrics to the CML.dev project that has some visual examples?

Reviewer 2: I already simplified it but good idea about CML.dev !

Reviewer 2: Applied in https://github.com/iterative/dvc.org/pull/1645#discussion_r463802976.

>   ML pipelines allow CI/CD systems to retrain
>   ([repro](https://dvc.org/doc/command-reference/repro)) models on fresh

Reviewer 2: ```suggestion
  ML pipelines allow CI/CD systems to retrain models on fresh
```

>   datasets with identical preprocessing and training stages, version upstream
>   models and datasets, and easily
>   [compare](https://dvc.org/doc/start/experiments#comparing-experiments) metrics
>   with currently deployed models.

Reviewer 2: Thanks but now we're introducing too many concepts here, I think. It's just an introductory doc 🙂 
```suggestion
  datasets with identical training, and save the results.
```
